client/post-editor/editor-gutenberg-dialogs/index.tsx

```diff
@@ -0,0 +1,59 @@
+/**
+ * External dependencies
+ */
+import React, { useState, useEffect } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+
+/**
+ * Internal dependencies
+ */
+import { showGutenbergOptInDialog } from 'state/ui/gutenberg-opt-in-dialog/actions';
+import EditorGutenbergOptInDialog from 'post-editor/editor-gutenberg-opt-in-dialog';
+import EditorGutenbergBlocksWarningDialog from 'post-editor/editor-gutenberg-blocks-warning-dialog';
+import { getEditorRawContent, isEditorLoading } from 'state/ui/editor/selectors';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import isSiteAutomatedTransfer from 'state/selectors/is-site-automated-transfer';
+import isPrivateSite from 'state/selectors/is-private-site';
+
+const EditorGutenbergDialogs: React.FC< {} > = () => {
+	const [ hasGutenbergBlocks, setHasGutenbergBlocks ] = useState< boolean | null >( null );
+
+	const siteId = useSelector( getSelectedSiteId ) as number;
+	const isAtomic = useSelector( state => isSiteAutomatedTransfer( state, siteId ) );
+	const isPrivate = useSelector( state => isPrivateSite( state, siteId ) );
+	const isPrivateAtomic = isAtomic && isPrivate;
+
+	const postContent = useSelector( getEditorRawContent );
+	const isLoading = useSelector( isEditorLoading );
+	const isPostContentLoaded = ! isLoading && postContent !== null;
+
+	const dispatch = useDispatch();
+
+	useEffect(
+		function() {
+			if ( hasGutenbergBlocks !== null || ! isPostContentLoaded ) {
+				return;
+			}
+			const hasGutenbergContent = ( postContent || '' ).indexOf( '<!-- wp:' ) !== -1;
+			setHasGutenbergBlocks( hasGutenbergContent );
+
+			// If a private atomic site uses classic editor, we want to always show some dialog, even
+			// if it doesn't have any Gutenberg blocks. Instead of saying "you may lose some data and formatting",
+			// let's just show the opt in dialog instead.
+			if ( ! hasGutenbergContent && isPrivateAtomic ) {
+				dispatch( showGutenbergOptInDialog() );
+			}
+		},
+		[ isPrivateAtomic, isPostContentLoaded, postContent ]
```

Comment:
Dependency array is missing `dispatch` and `hasGutenbergBlocks`

Reply:
> React guarantees that dispatch function identity is stable and won't change on re-renders. This is why it's safe to omit from the useEffect or useCallback dependency list.
https://reactjs.org/docs/hooks-reference.html#usereducer

Reply:
Regarding `hasGutenbergBlocks`, it's only set in that effect and it's meant to be only set once (if it's currently `null`) so it's missing on purpose - we don't want to rerun this effect after it sets `hasGutenbergBlocks`

Reply:
True. We're using the `dispatch` from `react-redux`, but it makes sense that they would make the same guarantee. Although this example from their docs includes it in the dependency list 🤔 
https://react-redux.js.org/next/api/hooks#examples

What initially alerted me to the issue was eslint, it's up to you whether you think it's better to just add the dependencies or add an eslint skip comment

Reply:
Yeah this should really be documented. Here's an info I found in a [similar discussion in another repo](https://github.com/reduxjs/redux-toolkit/issues/293):

> Usually, you do not need to, as dispatch will not change unless the react-redux Prodvider gets a new store and in a normal application that should never be the acse.
> But generally, it is considered best practice to just use the react eslint rules, and those do warn you if you don't add it in the dependencies array. So I'd just add it there, as it has no negative impact and you will follow best practices.

I'll just add it to dependencies

Reply:
And then I also had to add `// eslint-disable-next-line react-hooks/exhaustive-deps` to make lint OK with `hasGutenbergBlocks` missing

Reply:
>had to add // eslint-disable-next-line react-hooks/exhaustive-deps to make the linter OK with having hasGutenbergBlocks missing

Would be good to add note why it was disabled in case it's not obvious.